fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/mappers/MemoizingDimensionMappingResultSetMapper.java

```diff
@@ -0,0 +1,142 @@
+// Copyright 2019 Verizon Media Group
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.data.metric.mappers;
+
+import com.yahoo.bard.webservice.data.Result;
+import com.yahoo.bard.webservice.data.ResultSetSchema;
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.data.dimension.DimensionColumn;
+import com.yahoo.bard.webservice.data.dimension.DimensionField;
+import com.yahoo.bard.webservice.data.dimension.DimensionRow;
+
+import java.util.AbstractMap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A ResultSetMapper that transforms a ResultSet by transforming its {@link DimensionRow}s.
+ *
+ * If memoization is used, transformed rows will be cached.  The cost of memoization is proportional to the hashcode
+ * cost of the DimensionRow entries and the value is proportional to the number of repeated rows and the cost of
+ * transforming them.  In general, the expectation is for high redundancy of dimension rows per result set, so
+ * memoization is on by default.
+ */
+public class MemoizingDimensionMappingResultSetMapper extends ResultSetMapper {
+
+    private final BiFunction<DimensionColumn, DimensionRow, Boolean> columnRowMatcher;
```

Comment:
`java.util.function` has a `BiPredicate`: https://docs.oracle.com/javase/8/docs/api/java/util/function/BiPredicate.html that you could use. This should simplify this type some, and make it clearer what this is doing.

Reply:
done